crates/shared/src/price_estimation/trade_verifier.rs

```diff
@@ -184,9 +184,25 @@ impl TradeVerifier {
             }
         };
 
-        let summary = SettleOutput::decode(&output?, query.kind)
+        let mut summary = SettleOutput::decode(&output?, query.kind)
             .context("could not decode simulation output")
             .map_err(Error::SimulationFailed)?;
+
+        if verification.from == self.settlement.address() {
+            // Quote accuracy gets determined by how many tokens had to be paid out of the
+            // settlement buffers to make the quote happen. This does not work when the
+            // settlement contract itself is the trader.
+            // To not make it look like the entire trade was paid by the buffers we adjust
+            // the token balance differences based on the traded amounts.
+            let (sell_amount, buy_amount) = match query.kind {
+                OrderKind::Sell => (query.in_amount.get(), summary.out_amount),
+                OrderKind::Buy => (summary.out_amount, query.in_amount.get()),
+            };
+
+            summary.buy_tokens_diff += u256_to_big_rational(&buy_amount);
+            summary.sell_tokens_diff -= u256_to_big_rational(&sell_amount);
```

Comment:
Shouldn't this be the other way around? If the settlement contract sells token, at the end it will have a negative `sell_tokens_diff` which needs to be corrected by adding `sell_amount` (modelling the inflow of tokens).

Whereas, if the settlement contract is the receiver, we need to deduct the `buy_amount` (modelling the outflow in a normal trade).

Reply:
> Shouldn't this be the other way around? If the settlement contract sells token, at the end it will have a negative sell_tokens_diff which needs to be corrected by adding sell_amount (modelling the inflow of tokens).

The diff is computed by settlement contract balance `before - after` for sell and buy tokens. At the start the contract has a lot of sell tokens and afterwards only a few making the sell token diff overly positive.
For buy tokens we start with a low amount and end up with a high amount making the diff overly negative so we have to increase it by the buy token amount.

Reply:
But I guess computing the difference the other way around might be more intuitive. 😅